pipelines/zendesk/__init__.py

```diff
@@ -0,0 +1,253 @@
+    """
+    Dlt Resource that also writes to dlt state to update on custom ticket fields
+    @:param zendesk_client: Zenpy Object that is used to make API calls to Zendesk API
+    @:returns: Generator of dicts
+    """
+    # get dlt state
+    ticket_custom_fields = dlt.state().setdefault("ticket_custom_fields", {})
+    # get all custom fields and update state if needed, otherwise just load dicts into tables
+    all_fields = zendesk_client.ticket_fields()
+    for field in all_fields:
+        return_dict = field.to_dict()
+        field_id = str(field.id)
+        # grab id and update state dict if the id is new, add a new key to indicate that this is the initial value for title
+        if not (field_id in ticket_custom_fields):
+            ticket_custom_fields[field_id] = field.title
+            return_dict["initial_title"] = field.title
+        yield return_dict
+
+
+@dlt.resource(name="tickets", write_disposition="append")
+def ticket_table(zendesk_client: Zenpy, pivot_fields: bool = True, per_page: int = 1000, start_time: Optional[DateTime] = None) -> Iterator[TDataItem]:
+    """
+    Resource for tickets table. Uses DLT state to handle column renaming of custom fields to prevent changing the names of said columns.
+    This resource uses pagination, loading and side loading to make API calls more efficient
+    @:param zendesk_client: Zenpy type object, used to make calls to Zendesk API through zenpy module
+    @:param pivot: bool that indicates whether to pivot the custom fields in tickets, default set to true
+    @:param per_page: Describes how many Ticket objects are loaded per page, int
+    @:parm start_time: The datetime for which Tickets start to load, tickets before this date will not be loaded
+    @:returns: Generator returning dicts for every row of data
+    """
+
+    # grab the start time from last dlt load if not filled, if that is also empty then use the first day of the millennium as the start time instead
+    if not start_time:
+        start_time = dlt.state().setdefault("last_load_tickets", FIRST_DAY_OF_MILLENNIUM.timestamp())
+    else:
+        # if start time is given convert to a timestamp as it is required by the zenpy api client
+        start_time = start_time.timestamp()
+    # grab the custom fields from dlt state if any
+    fields_dict = dlt.state().setdefault("ticket_custom_fields", {})
+    all_tickets = zendesk_client.tickets.incremental(paginate_by_time=False,
+                                                     per_page=per_page,
+                                                     start_time=start_time,
+                                                     include=["users", "groups", "organisation", "brands"])
+    for ticket in all_tickets:
+        ticket_dict = process_ticket(ticket=ticket, custom_fields=fields_dict, pivot_fields=pivot_fields)
+        yield ticket_dict
+    # update the last load time in the dlt state
+    dlt.state()["last_load_tickets"] = now().timestamp()
```

Comment:
is this really now()? why not the last timestamp in the ticket? will it return duplicate tickets? even if so you have better chances of not losing anything if you add a milisecond to the last timestamp. or you just remove the last ticket in the list so you'll get it next time

Reply:
The way I thought about this would be the most straightforward way of saving the last load time. Perhaps it is more accurate to save the timestamp when the query is initially ran and save that as the last load time so we don't lose any tickets created while the pipeline is running. Checking every ticket for the timestamps and updating to the latest one seems unnecessarily complex.

Reply:
this is how it is done though. we want to automate it btw, https://github.com/dlt-hub/dlt/issues/194
you need to put the last timestamp from the ticket in there.
then you need to check if you are getting any duplicates. I suggested a test to do that